### PR TITLE
ENH: stats: add 'alternative' keyword to some normality tests.

### PR DESCRIPTION
**Pull request by tirthasheshpatel**
#### Reference issue

Addresses gh-12506

#### What does this implement/fix?

This PR adds `alternative` parameter to `skewtest`, `kurtosistest`,
and `ranksums` functions to allow one-tailed hypothesis testing. An example
elucidating the use of this parameter has been added in the documentation
suite of each of the changed functions. Necessary tests to ensure full
coverage have also been added.

A bit unrelated, but necessary change in the tests of the `ranksums`
function has also been made. The tests of `ranksums` did not validate
the outputs previously. Hence, a more comprehensive testsuite, comparing
the outputs with R's `wilcox.test`, has been added.

Other normality tests (present in the table gh-12506), namely `kendalltau`,
`ansari`, and `mood`, are a bit tricky to handle. Hence, I have left them
unchanged for now.

### CODE REVIEW

**Comment by tirthasheshpatel** [comment]
The `scipy.scipy (Windows Python38-64bit-full-ilp64)` failure is unrelated. Refguide check failure is due to a small floating point error which is dependent on the system on which it is run. Should I remove that output from the doc suite or is there some other fix for it?

**Comment by mdhaber** [comment]
On the line with the output, add `# may vary`.

**Comment by mdhaber** [comment]
Looks like a great start! Can't do full review now, but one thing I was hoping to add with these is the meaning of the alternatives, as it is not always obvious what is "greater" than what. Precedent doesn't require it, but would you be willing to note that in the documentation?

**Comment by tirthasheshpatel** [comment]
> Looks like a great start! Can't do full review now, but one thing I was hoping to add with these is the meaning of the alternatives, as it is not always obvious what is "greater" than what. Precedent doesn't require it, but would you be willing to note that in the documentation?

Sure! I will see what I can add to explain the meaning/use of the `'less'` and `'greater'` values that the `alternative` parameter can take.

File: scipy/stats/stats.py
Comment (mdhaber) [inline scipy/stats/stats.py:1496]: To anyone else who happens to see - any preference between `z[()]` and `z.item()`?

File: scipy/stats/stats.py
Comment (mdhaber) [inline scipy/stats/stats.py:7571]: So in this function, which did not do something like
```python3
    if z.ndim == 0:
        z = z[()]
```
are we changing the behavior? I imagine that this is an unusual case; just thought I should check.

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:4455]: Not a big deal, but you could do away with these assignments. I don't recall running across any statistics that change based on the alternative being compared, so I don't think just comparing against `st_skew` would be too confusing.

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:4551]: Is it not accurate to more digits? Do you know why?

File: scipy/stats/stats.py
Comment (tirthasheshpatel) [reply scipy/stats/stats.py:7571]: > are we changing the behavior?

No, we aren't changing the behaviour here. It returned a `np.float64` for 1-D inputs before still does the same. Apparently, doing `z[()]` returns a `np.float64` for `np.float64` or `ndarray` `z` but `z.item()` returns a python float! Maybe, that's one of the reasons scipy functions prefer doing `z[()]` over `z.item()`.

File: scipy/stats/tests/test_stats.py
Comment (tirthasheshpatel) [reply scipy/stats/tests/test_stats.py:4551]: > Is it not accurate to more digits? Do you know why?

Unfortunately, they only match up to 2 digits and I don't know why.

For reference, this is the code I am using in R:

```R
options(digits=15)

women_weight <- c(38.9, 61.2, 73.3, 21.8, 63.4, 64.6, 48.4, 48.8, 48.5)
men_weight <- c(67.8, 60, 63.4, 76, 89.4, 73.3, 67.3, 61.3, 62.4)

wilcox.test(women_weight, men_weight, exact = FALSE, correct = FALSE)
```

and this is the output:

```none

	Wilcoxon rank sum test

data:  women_weight and men_weight
W = 15, p-value = 0.024194132827
alternative hypothesis: true location shift is not equal to 0
```

Am I doing something wrong?

File: scipy/stats/tests/test_stats.py
Comment (tirthasheshpatel) [reply scipy/stats/tests/test_stats.py]: Done!

File: scipy/stats/stats.py
Comment (mdhaber) [reply scipy/stats/stats.py:7571]: Thanks.

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [reply scipy/stats/tests/test_stats.py:4549]: I don't think so. `mannwhitneyu` matches R, though.
```python3
import numpy as np
from scipy.stats import mannwhitneyu, ranksums
women_weight = [38.9, 61.2, 73.3, 21.8, 63.4, 64.6, 48.4, 48.8, 48.5]
men_weight = [67.8, 60, 63.4, 76, 89.4, 73.3, 67.3, 61.3, 62.4]
pvalue = 0.024194132827
res1 = ranksums(women_weight, men_weight)
res2 = mannwhitneyu(women_weight, men_weight, alternative='two-sided', use_continuity=False)
np.testing.assert_allclose(res2.pvalue, pvalue)
print(res1.pvalue - res2.pvalue)
```
Would you be willing to look into what mannwhitneyu is doing differently?

Comment (tirthasheshpatel) [reply scipy/stats/tests/test_stats.py:4549]: > Would you be willing to look into what mannwhitneyu is doing differently?

Yeah, let me check. I will let you know if I see something interesting.

Comment (tirthasheshpatel) [reply scipy/stats/tests/test_stats.py:4549]: Ok, so I found something that makes sense. We see the difference between `ranksums` and `mannwhitneyu` because `mannwhitneyu` corrects for ties while `ranksums` doesn't. In the line below:

https://github.com/scipy/scipy/blob/4fe52aed1ea375c2aa7a2a6b634787c9abdf32e3/scipy/stats/stats.py#L7417

replace `T` (here `T` is the correction for ties. see line 7414) with `1` so it becomes: `sd = np.sqrt(1 * n1 * n2 * (n1+n2+1) / 12.0)`

and then the code that you provided gives this output:

```python
>>> import numpy as np
>>> from scipy.stats import mannwhitneyu, ranksums
>>> women_weight = [38.9, 61.2, 73.3, 21.8, 63.4, 64.6, 48.4, 48.8, 48.5]
>>> men_weight = [67.8, 60, 63.4, 76, 89.4, 73.3, 67.3, 61.3, 62.4]
>>> pvalue = 0.024194132827
>>> res1 = ranksums(women_weight, men_weight)
>>> res2 = mannwhitneyu(women_weight, men_weight, alternative='two-sided', use_continuity=False)
>>> # np.testing.assert_allclose(res2.pvalue, pvalue)
>>> print(res1.pvalue - res2.pvalue)
0.0
```

I don't think there is a way to avoid correction for ties in R's `wilcox.test` so we should probably use the uncorrected value `res2.pvalue` in the code snippet above. What do you say?

Comment (mdhaber) [reply scipy/stats/tests/test_stats.py:4549]: Ah, of course. That would work, but I would prefer that it's something that can be verified without changing `mannwhitneyu` code. What do you think of using `mannwhitneyu` to calculate the "ground truth" value of the statistic, then standardizing it and getting the p-value from `norm` right there in the test? Formula is pretty simple, especially without ties.

Comment (tirthasheshpatel) [reply scipy/stats/tests/test_stats.py]: Done. Thanks!

**Comment by mdhaber** [comment]
Great! Please ping me when the documentation is ready.

**Comment by tirthasheshpatel** [comment]
> Great! Please ping me when the documentation is ready.

Done, @mdhaber. I have tried to write more informative docs for the `alternative` parameter.

File: scipy/stats/stats.py
Comment (mdhaber) [inline scipy/stats/stats.py:1541]: This is better. 
Just as we define Z as the observed value of the test statistic, z would need to be defined as the test statistic under the null hypothesis. 
Ideally, wouldn't we also tell the user what the test statistic for this test is? Would they know without looking at the source code?
On the other hand, I'm not sure we want to write it in such a way that the reader needs to be familiar with the z score test statistic. For instance, in the t-test, we would write about the difference in means (not the t statistic), and in kstest we write about the difference in CDF (not the precise statistic).
@chrisb83 how do you think we should document what alternatives mean, in general?

**Review by mdhaber — COMMENTED** [review]
I'll take another look at the tests at some point and will consider merging as is. The documentation of the alternatives is specific enough; perhaps improvements can happen after a bigger discussion about how to document alternatives.

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:4472]: The tests seem right. 

I suppose if i think about it, I can see that the skewness of `x = np.array((-2, -1, 0, 1, 2, 3)*4)**2` should be positive. It is not obvious to me that I would expect the kurtosis to be less than that of the normal distribution. In any case, I wish the original tests were more obvious.

For example:
```python3
a = skewnorm.rvs(a=1, size = 10000)  # skew parameter is 1 > 0
print(skewtest(a, alternative='greater'))
```
should have a small p-value

and

```python3
a = laplace.rvs(size = 1000)  # excess kurtosis of laplace is 3 > 0
print(kurtosistest(a, alternative='greater'))
```
should have a small p-value.

But I checked these locally and they looks good, so I leave it up to you whether you want to add them. I suppose this is more of a suggestion for the original authors than this PR.

**Review by mdhaber — COMMENTED** [review]
Some final thoughts. Everything is non-blocking except that lower case z needs to be defined. We can leave the rest as they are, as they seem correct and overall this is an improvement. I just think it would be nice if we would clean up the documentation and testing of these old functions while our attention is on them.

File: scipy/stats/stats.py
Comment (tirthasheshpatel) [reply scipy/stats/stats.py:1541]: > If you agree, I am not sure that we should leave this sort of thing for future work. The more I work with stats, the more I think that if we're going to make improvements like this, we should do the full job.

I agree with point 2). Yeah, we can leave this idea to a future PR that cleans up all the documentation in stats. For now, I will just define what z is. Is that ok?

File: scipy/stats/tests/test_stats.py
Comment (tirthasheshpatel) [reply scipy/stats/tests/test_stats.py:4472]: You are right. I will add the tests that are easier to understand. Thanks for the suggestion.

File: scipy/stats/stats.py
Comment (mdhaber) [reply scipy/stats/stats.py:1541]: Thanks. Sorry it's not as simple as you originally planned! But I think it will be much more useful when you're done.

Comment (tirthasheshpatel) [reply scipy/stats/stats.py]: > Thanks. Sorry it's not as simple as you originally planned! But I think it will be much more useful when you're done.

No worries, @mdhaber. I have attempted to change the documentation to be more intuitive. And yes, it does sound way more useful than it originally did. Even R prints the alternative hypothesis when a test is run. So, this change is definitely worth it.

File: scipy/stats/tests/test_stats.py
Comment (tirthasheshpatel) [reply scipy/stats/tests/test_stats.py:4472]: The tests you have suggested here have been added now.

**Review by mdhaber — APPROVED** [review]
This is an improvement. The documentation is clear, and although there might be other improvements to make (e.g. formatting - indentation of bulleted lists causes them to render as a quotation), it is better than most stats functions, which don't describe the alternatives at all. I think the definitions (in the code) and descriptions of the alternative hypotheses are consistent and correct. I think I'll give this another few days for comments, and I'll merge Sunday if there are none. Thanks @tirthasheshpatel!

**Comment by mdhaber** [comment]
@DominicChm would you review this, since you've added some `alternatives` yourself? That is, pretending I hadn't reviewed this, could you check for accuracy of the descriptions of the alternatives and consistency between the direction of the description and the way it's defined in code?

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:4484]: In the future, let's use `assert_allclose` pretty much always:

image

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:4553]: @tirthasheshpatel It seems there was no need to use that data, since we didn't have an external p-value to compare against. In that case, why not just use something without ties so that the p-values would be the same?

Do you agree that this does the same job?

Also, splitting things into separate functions, each with a particular job, makes it a little easier to read IMO.

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:4565]: I think it's important to check the message. There are a lot of `ValueErrors`, so it's good to check for the one you expect.

File: scipy/stats/stats.py
Comment (mdhaber) [inline scipy/stats/stats.py:7547]: Which way is the location shift defined? `x - y` or `y - x`? I think "location shift" makes some additional assumptions about the test, so I'm going to suggest using the idea of "stochastically greater/less".

File: scipy/stats/tests/test_stats.py
Comment (tirthasheshpatel) [reply scipy/stats/tests/test_stats.py:4553]: > Do you agree that this does the same job?

Yes, this should be equivalent to what we had before + this is simpler.

> Also, splitting things into separate functions, each with a particular job, makes it a little easier to read IMO. Do you agree?

This makes sense. It's more readable to me too. Thanks!

File: scipy/stats/stats.py
Comment (mdhaber) [inline scipy/stats/stats.py:7542]: Let's not add indent, because that adds the vertical bars.

File: scipy/stats/tests/test_stats.py
Comment (tirthasheshpatel) [reply scipy/stats/tests/test_stats.py:4565]: > I think it's important to check the message. There are a lot of `ValueErrors`, so it's good to check for the one you expect.

Yes. This is better. Thank you!

File: scipy/stats/tests/test_stats.py
Comment (tirthasheshpatel) [reply scipy/stats/tests/test_stats.py:4484]: Thanks for the suggestion. I will keep this in mind from next time onwards.

File: scipy/stats/stats.py
Comment (tirthasheshpatel) [reply scipy/stats/stats.py]: > Which way is the location shift defined? `x - y` or `y - x`?

It's `x - y`. 

> I think "location shift" makes some additional assumptions about the test, so I'm going to suggest using the idea of "stochastically greater/less".

Sure, I will change it to that.

Comment (tirthasheshpatel) [reply scipy/stats/stats.py]: Oh, I see you just changed it. Thanks!

Comment (mdhaber) [reply scipy/stats/stats.py]: Is this new wording OK? Do you agree that it corresponds with the [more general formulation](https://en.wikipedia.org/wiki/Mann%E2%80%93Whitney_U_test#:~:text=Under%20the%20general)?

File: scipy/stats/stats.py
Comment (mdhaber) [reply scipy/stats/stats.py:7758]: Here's another reason: z[()] works on pretty much anything NumPy returns in just the way you'd want it to. If it's already a float, no problem. If it's a 0d array, it returns a float. If it's any other array, it just returns the array.

**Review by mdhaber — APPROVED** [review]
If the recent changes look good to you @tirthasheshpatel; I'll merge when CI passes (if the documentation renders as intended).

File: scipy/stats/stats.py
Comment (tirthasheshpatel) [reply scipy/stats/stats.py]: > Is this new wording OK? Do you agree that it corresponds with the [more general formulation](https://en.wikipedia.org/wiki/Mann%E2%80%93Whitney_U_test#:~:text=Under%20the%20general)?

Yes, I think the new description looks correct.

File: scipy/stats/stats.py
Comment (tirthasheshpatel) [inline scipy/stats/stats.py:1538]: I think that explicitly stating that this is the alternative hypothesis is better here. 

```suggestion
        * 'two-sided': The alternative hypothesis is that the skewness of the distribution underlying the sample
```

From the view of a new user, this makes it easier to understand whether the statement is the null hypothesis or the alternative hypothesis. Also, IIRC, other tests that you have documented do this. Do you think doing this would be helpful here too?

Comment (mdhaber) [reply scipy/stats/stats.py:1538]: > Also, IIRC, other tests that you have documented do this. 

Do you mean gh-13188? I was just pushing buttons for others there. The reason why it says "the alternative is..." every time is that it first says "the null hypothesis is ..." for each option. But in that PR, I don't understand how you have different null hypotheses corresponding with the different alternatives. For instance, I understand

> The null hypothesis is that F(x) = G(x) for all x

But how do you calculate the distribution of the test statistic under

> The null hypothesis is that F(x) >= G(x) for all x

? Is that really what's done? 

In every statistical test I've thought much about, there is just one null hypothesis corresponding for all three alternative hypotheses. The thing that changes is how you calculate the p-value given the distribution of the test statistic under the assumption of the null hypothesis.

---

Back to this PR. If we're not stating the null hypothesis, I don't think it needs to be stated so many times that this is the alternative hypothesis. I wanted to remove that redundancy. 

I thought about moving it above like:
> Defines the alternative hypothesis. Default is 'two-sided'.
   The following options *for the alternative hypothesis* are available:

But that's still redundant. The parameter is called "alternative", and we just said "Defines the alternative hypothesis", so I would be surprised if someone still wondered whether these were null hypotheses or alternative hypotheses. We said it was the alternative twice already.

On the other hand, I do agree with the idea that we should state the null hypothesis. So if we were to do that, I might write:

> Defines the alternative hypothesis. Default is 'two-sided'.

> The null hypothesis is that the skewness of the distribution underlying the sample is the same as that of the normal 
   distribution. The following options for the alternative hypothesis are available:

Would that be better? 
Is that accurate? Is it true that the null hypothesis is the same for all three cases?

Comment (mdhaber) [reply scipy/stats/stats.py:1538]: Never mind. The null is stated at the top. I say leave it there.

Comment (tirthasheshpatel) [reply scipy/stats/stats.py:1538]: > Do you mean gh-13188?

Yes, I was talking about that PR.

> Would that be better?

I think it would be OK both ways: writing null and alternative hypotheses or just writing the alternative hypothesis. For the sake of completeness, it would be better to write both the null and alternate hypotheses.

> Is that accurate? Is it true that the null hypothesis is the same for all three cases?

I don't think so. For instance, when `alternative='less'` for the skew test, the null hypothesis is that the skewness of the sample is greater than that of the normal distribution (i. e. > 0). It also changes the same way for when `alternative='greater'`. If this sounds like too much work for this PR, we can leave this for some other time. What do you say?

File: scipy/stats/stats.py
Comment (mdhaber) [inline scipy/stats/stats.py:1645]: At the top, it says:

image

I don't understand. What does `3*(n+1)/(n-1)` refer to? 

_Assuming_ `n` is the number of observations in the sample `a` (because it is not stated): 
* the kurtosis of a _population_ does not depend on the number of observations of a sample drawn from it
* what is this formula? We use Fisher's definition of kurtosis, so the kurtosis of the normal distribution is 0, right?

Should I just remove this formula?

File: scipy/stats/stats.py
Comment (tirthasheshpatel) [reply scipy/stats/stats.py:1538]: > Do you still think these might be confused with null hypotheses?

I don't think it engenders confusion. I thought that being explicit would just make it easier to understand. But from your comment earlier:

> But that's still redundant. The parameter is called "alternative", and we just said "Defines the alternative hypothesis", so I would be surprised if someone still wondered whether these were null hypotheses or alternative hypotheses. We said it was the alternative twice already.

I think it's OK not to introduce redundancy.

Comment (mdhaber) [reply scipy/stats/stats.py:1538]: OK. Thanks for the rapid feedback! So much easier to get things done in near real-time!

Comment (mdhaber) [reply scipy/stats/stats.py:1538]: But

> I don't think so. For instance, when alternative='less' for the skew test, the null hypothesis is that the skewness of the sample is greater than that of the normal distribution (i. e. > 0)

We're back to something I don't understand, as I wrote [above regarding kstest](https://github.com/scipy/scipy/pull/13549#discussion_r594923594). If this is the case, then the null hypothesis at the top of the documentation is not correct for all three alternatives and would need to be changed.

Comment (mdhaber) [reply scipy/stats/stats.py:1538]: Consider, for example, [the binomial test](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.binom_test.html).

The null hypothesis for all three alternatives is the same: that the probability of "success" is the provided `p`. The distribution of the test statstic (the obverved number of successes) is calculated under that same assumption for all alternatives. The alternatives just control which tail you look at. Right?

How are these other tests different?

Comment (tirthasheshpatel) [reply scipy/stats/stats.py:1538]: > Consider, for example, [the binomial test](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.binom_test.html).
> 
> The null hypothesis for all three alternatives is the same: that the probability of "success" is equal to the provided `p`. The distribution of the test statstic (the observed number of successes) is calculated under that same null hypotheis for all alternatives - it is the binimial distribution with the provided `n` and `p`. The alternatives just control which tail you look at. Right?
> 
> How are these other tests different?

I think that the null hypothesis is everything other than the alternate hypothesis for normality tests. I will have to see if that still holds true for other tests though. If I am not wrong, we cannot say if the statistic is exactly the null value if we get a high p-value at the left/right tail which is why we formulate the null hypothesis in such a way that covers the rest of the distribution (which is not covered by the alternate hypothesis). Does this make sense? I will try to rephrase or attach some source if it's unclear.

File: scipy/stats/stats.py
Comment (tirthasheshpatel) [reply scipy/stats/stats.py:1645]: > the kurtosis of a population does not depend on the number of observations in a sample drawn from it, right?

Right.

> We use Fisher's definition of kurtosis, so the kurtosis of the normal distribution is 0, right?

Right.

> Should I just remove this formula?

Sure, it sounds like the right thing to do!

Comment (mdhaber) [reply scipy/stats/stats.py:1645]: OK, I see from the reference that this is the mean of the distribution of the sample kurtosis (not the excess/Fisher kurtosis) when the sample is drawn from the normal distribution. Using git blame, I see that the original docstring was:
> Tests whether a dataset has normal kurtosis (i.e., kurtosis=3(n-1)/(n+1)).

This makes some sense, as it's a statement about the dataset, and perhaps SciPy's convention of using excess kurtosis was not as fully-formed then.

But given the rest of current docstring, I think it should be removed.

Comment (tirthasheshpatel) [reply scipy/stats/stats.py:1645]: I agree. Thanks for noticing this. I am out right now so will push this change when I am back home. I guess this is ready to merge afterwards (unless someone else reviews this), right?

File: scipy/stats/stats.py
Comment (mdhaber) [reply scipy/stats/stats.py:1538]: > If I am not wrong, we cannot say if the statistic is exactly the null value

Statistical tests never tell us whether the null hypothesis is true. They can only give us evidence for rejecting it. And that distinction is _very_ important to this issue. 

So I'm afraid that doesn't make sense as written. I wrote a longer post explaining, but I'll wait until you reply to post it.

File: scipy/stats/stats.py
Comment (mdhaber) [reply scipy/stats/stats.py]: I went ahead. If it looks good to you, I'll merge.

Comment (mdhaber) [reply scipy/stats/stats.py]: I also removed "(i.e. 0)" to avoid having to explain the different conventions for kurtosis.

Comment (tirthasheshpatel) [reply scipy/stats/stats.py]: Everything looks good to me. So +1 for merging. (My comment about being explicit about the alternate hypothesis is non blocking. So, please don't let it bother you from merging. Seems like I have got something wrong there... )

File: scipy/stats/stats.py
Comment (tirthasheshpatel) [reply scipy/stats/stats.py:1538]: > Statistical tests never tell us whether the null hypothesis is true. They can only give us evidence for rejecting it. And that distinction is _very_ important to this issue.

That aligns with my understanding. I wasn't able to write it properly. Sorry!

> So I'm afraid that doesn't make sense as written. I wrote a longer post explaining, but I'll wait until you reply to post it. You might want to read the conversation at [gh-12518](https://github.com/scipy/scipy/issues/12518), as this has been discussed before.

I looked at gh-12518. My understanding about the null hypothesis being equality/weak inequality was wrong. Statsmodels also uses the equality null hypothesis. So, it would make sense to do that here too. Very sorry for the noise here. Please go ahead and post your longer post if you want. Also, thanks for writing a long post! Really appreciate your help here!

Comment (mdhaber) [reply scipy/stats/stats.py:1538]: OK here it is.

---
If failing to reject the null hypothesis meant "The null hypothesis is correct", e.g. "the true value of the statistic is exactly the null value", then I can see why we might have a problem stating the null hypothesis with a precise value. But there are lots of other problems in that alternate universe : )

Here, we 
1. assume a (precise) null hypothesis about the underlying population,
2. derive a distribution of the sample statistic under the assumption of the null hypothesis, and
3. compare the observed value of the statistic against that distribution. 

If the observed value of the statistic is sufficiently extreme when compared to the disribution (where "sufficiently extreme" is defined by our choice of alternative and significance level), then we say we have evidence to reject the null hypothesis in favor of the alternative. When we fail to reject the null hypothesis, we do not confirm that the null hypothesis is true. We just don't have evidence in favor of the alternative.

Regarding the tests in this PR: the only thing that changes in these tests for the different alternatives is the way we compare the observed value of the test statistic against the distribution (which is derived under the null hypothesis). So even from the code we can see that distribution - and thus the null hypothesis itself - doesn't change depending on the choice of alternative.

**Comment by mdhaber** [comment]
Thanks @tirthasheshpatel!

**Pull request merged**
